Support node maintenance mode

dotnwat opened the pull request:
## Cover letter

Adds three administrative interfaces:

```
   Start maintenance:  PUT    /v1/brokers/ID/maintenance
   Stop maintenance:   DELETE /v1/brokers/ID/maintenance
   Drain status: GET    /v1/brokers/ID/maintenance
```

which allow a node to be placed into a maintenance mode in which the primary action is to relinquished all leadership from a node. if a node is taken out of the maintenance state then leadership is allowed to return. this is a building block for rolling upgrades in which a target node to upgrade is drained to minimize disruption to client traffic.

maintenance mode is persistent and stored in the controller. currently, only one node at a time may be in maintenance mode.

reviewers may notice some obvious failure scenarios. for example, if leadership cannot be transferred off of a node then draining can never complete. this is by design: initially the design assumes that an external process (e.g. k8s operator) will monitor cluster health and draining progress. it may choose to proceed with upgrade despite some leadership failing to transfer, or may take a node out of draining state to deal with the underlying issue.

this feature and the heuristics for dealing with cluster situations is sure to expand. this initial set of functionality should be sufficient to allow k8s operator work to proceed.

Fixes: #3706
Fixes: #3705

## Release notes

* Support for placing node into a draining state in which all leadership is relinquished.

dotnwat commented:
@simon0191 up in the cover letter are the administration control URLs we discussed yesterday in the context of rolling upgrades.

mmaslankaprv commented on src/v/redpanda/drain_manager.cc:
```diff
@@ -0,0 +1,267 @@
+        size_t failed = 0;
+        for (auto& f : results) {
+            try {
+                auto err = f.get0();
```
nit: use `get()`

dotnwat replied:
Done

mmaslankaprv commented on src/v/redpanda/drain_manager.cc:
```diff
@@ -0,0 +1,267 @@
+         */
+        auto dur = ss::lowres_clock::now() - started;
+        if (failed > 0 && dur < transfer_throttle && _draining && !_stop) {
+            co_await ss::sleep(transfer_throttle - dur);
```
abortable ?

dotnwat replied:
Done

mmaslankaprv commented:
This looks great. only two minor nits

jcsp commented on src/v/raft/consensus.cc:
```diff
@@ -2439,6 +2439,21 @@ ss::future<timeout_now_reply> consensus::timeout_now(timeout_now_request&& r) {
+    if (_node_priority_override == zero_voter_priority) {
+        vlog(
+          _ctxlog.debug,
+          "Ignorning timeout request in state {} with node voter priority zero "
```
s/Ignorning/Ignoring/

dotnwat replied:
Fixed

jcsp commented:
High level thoughts:
- Is draining==maintenance mode?  The latter is probably a more idiomatic name for the state.  Draining is also prone to misunderstanding because it sounds like we might be draining the data from the node, as opposed to just leadership.
- We might have more nodes in future (maybe to distinguish a node which is draining for restart vs. a node draining for decom?) maybe the API should be a bit more extensible, for example rather than cancelling drain mode with a DELETE to /v1/drain, it could be a PUT {"status": "active"} to /v1/node/<thisnode>.
- To be really safe, the state should be persistent -- if it's ephemeral, then the operator has no guarantee that the node won't restart and take up leaderships while it is supposed to be in the draining state.
- In future, it would be ideal for some central point to know which nodes are in the draining state.  That could be useful for situational awareness but also for restricting when a node can be put into the mode -- a central point of control could enforce rules preventing too many nodes being in maintenance at once.

I guess a couple of those points would be covered if we drove the mode via a state in members_table -- that's a bit more invasive of course (needs new structs/encoding & a feature flag) but probably where we want to be longer term?

senior7515 commented on src/v/cluster/partition_manager.cc:
```diff
@@ -89,6 +89,19 @@ ss::future<consensus_ptr> partition_manager::manage(
+     * synchronization with the drianing manager.
+     */
+    if (_block_new_leadership) {
+        p->block_new_leadership();
```
should this have a timeout to unblock, say 30mins or 1hr or smth in case of a cluster split-brain after the draining started and can't make progress.

senior7515 commented:
@dotnwat should draining have a timeout like logging where `-1` is infinite as part of the API - i.e.: the node should drain in 3 hrs or something terrible has happened kinda thing? just a thought.

BenPope commented on src/v/redpanda/admin_server.h:
```diff
@@ -14,6 +14,7 @@
 #include "config/endpoint_tls_config.h"
 #include "coproc/partition_manager.h"
 #include "model/metadata.h"
+#include "redpanda/drain_manager.h"
```
Only a forward declaration is required (which true of some of the other includes, too) - this is included from `redpanda/application.h`

dotnwat replied:
fixed

BenPope reviewed (commented):
Very nice.

dotnwat commented on src/v/raft/consensus.cc:
```diff
@@ -2439,6 +2439,21 @@ ss::future<timeout_now_reply> consensus::timeout_now(timeout_now_request&& r) {
+        vlog(
+          _ctxlog.debug,
+          "Ignorning timeout request in state {} with node voter priority zero "
+          "from {} at term {}",
```
fixed

dotnwat commented on src/v/cluster/partition_manager.cc:
```diff
@@ -89,6 +89,19 @@ ss::future<consensus_ptr> partition_manager::manage(
+     * synchronization with the drianing manager.
+     */
+    if (_block_new_leadership) {
+        p->block_new_leadership();
```
good point. I fixed this

dotnwat commented on src/v/redpanda/application.cc:
```diff
@@ -1026,6 +1026,10 @@ void application::wire_up_redpanda_services() {
         priority_manager::local().compaction_priority()))
       .get();
+
+    construct_service(
+      drain_manager, std::ref(_log), std::ref(partition_manager))
+      .get();
```
mm, I'm not sure it matters because when the drain manager is stopped there is are no state changes (e.g. to blocking leadership) that occur.

dotnwat replied:
became irrelevant because I moved this into the controller.

dotnwat commented on src/v/redpanda/drain_manager.cc:
```diff
@@ -0,0 +1,267 @@
+     */
+    _partition_manager.local().block_new_leadership();
+
+    for (const auto& p : _partition_manager.local().partitions()) {
+        p.second->block_new_leadership();
+    }
```
Nice idea. Done.

dotnwat commented on src/v/redpanda/drain_manager.cc:
```diff
@@ -0,0 +1,267 @@
+
+    _partition_manager.local().unblock_new_leadership();
+
+    for (const auto& p : _partition_manager.local().partitions()) {
+        p.second->unblock_new_leadership();
+    }
```
Nice idea. Done.

dotnwat commented:
> How does this interact with decommissioning?
> Is draining a part of decommission? Should it be?

I'm under the impression that decommissioning affectively drains, but instead of draining leadership it actually moves partitions off the node.

> What happens if the drain API is called on a decommissioned node?

Mm, I don't think anything would happen. If it's decommissioned it can be removed.

> Would it make sense to wire this into application::shutdown?

Yes! The plan was to do this in a follow up PR after the basic drain mechanism is robust.

dotnwat commented:
Notable changes in latest iteration (cc @jcsp @mmaslankaprv @BenPope)

1. Rename from drain -> maintenance mode. Internally we still use a drain manager for detailing with draining leadership, but other mode and external interfaces are expressed in terms of a maintenance mode.

2. Maintenance mode is now persistent and stored in the members_table / raft0. The current policy is to only allow one node at a time to be in maintenance mode. This should allow us to build some tooling for bare-metal, too, via rpk.

Feedback from @jcsp that wasn't done:

1. I agree with making the interface more extensible by generalizing the representation of node status. I got stuck a bit in design space land with this. Like, should we unify several things like decommission state, etc...? I'm not sure but it seems like something we can clean up at some point.

One annoying thing, that is probably ok for now:

1. Clients are expected to set maintenance mode via normal interface that gets redirected to controller and made persistent. But clients should follow up after this by querying the target node for when maintenance mode has been fully entered.
2. This decision made the protocol much simpler by not having to build a state machine between the controller and each node which would track the phases of maintenance mode.
3. This can probably be deprecated at some point and fixed up transparently in rpk and other tools if we decide to track the life cycle explicitly in the controller.

What should come in follow up PRs

1. Blocking traffic that is interfering with progress on leadership transfer. This should in principle be straight forward to implement assuming a fairly simple heuristic.

mmaslankaprv commented:
The code looks good, however there is an assertion triggered in one of the tests:
```
rptest.services.redpanda.BadLogLines: <BadLogLines nodes=docker_n_4(4) example="ERROR 2022-03-10 08:10:12,917 [shard 0] assert - Assert failure: (../../../src/v/cluster/drain_manager.cc:26) '_drain.has_value()' service has not been started">
```

BenPope commented on src/v/cluster/controller.cc:
```diff
@@ -197,6 +200,8 @@ ss::future<> controller::start() {
             std::ref(_storage),
             std::ref(_as));
       })
+      .then(
+        [this] { return _drain_manager.invoke_on_all(&drain_manager::start); })
```
I don't see an equivalent for `&drain::manager::stop`

dotnwat replied:
the service is stopped via `[this] { return _drain_manager.stop(); }` later on in the file. is that what you mean?

BenPope replied:
Oh right, I forgot that knew how to stop the underlying service.

dotnwat replied:
> Oh right, I forgot that knew how to stop the underlying service.

I find this asymmetry with the api kind of annoying too

BenPope commented on src/v/cluster/drain_manager.h:
```diff
@@ -0,0 +1,93 @@
+    ss::future<> start();
+    ss::future<> stop();
+
+    /*
+     * Start draining this broker.
+     *
+     * Invoke this on each core.
+     */
+    ss::future<> drain();
+
+    /*
+     * Restore broker to a non-drain[ing] state.
+     *
+     * Invoke this on each core.
+     */
+    ss::future<> restore();
```
Any reason not to encapsulate the invocation on every core?

dotnwat replied:
hmm, not exactly. just trying to write things so it would be easier to unit test without sticking it in a sharded<T>. but I think I messed that up because the status() does a global map_reduce. probably should have made that a free function :/

BenPope reviewed (commented):
Looks great,

I think there should be a few more tests

dotnwat commented:
> The code looks good, however there is an assertion triggered in one of the tests:
> 
> ```
> rptest.services.redpanda.BadLogLines: <BadLogLines nodes=docker_n_4(4) example="ERROR 2022-03-10 08:10:12,917 [shard 0] assert - Assert failure: (../../../src/v/cluster/drain_manager.cc:26) '_drain.has_value()' service has not been started">
> ```

Oh thanks. Yeh there is some shutdown path I missed I guess.

dotnwat commented:
> Looks great,
> 
> I think there should be a few more tests

I will do this before tomorrow. I totally agree

BenPope commented:
> > Looks great,
> > 
> > I think there should be a few more tests
> 
> I will do this before tomorrow. I totally agree

I wouldn't necessarily let it block a release. I'm mostly concerned with the failure paths.

dotnwat commented:
* Resolved feedback and fixed CI failure
* Added feature flag and bumped cluster version
* Added tests
  * Verify that maintenance mode is persistent
  * Verify that one-at-a-time-rule is enforced

mmaslankaprv commented on src/v/redpanda/admin_server.cc:
```diff
@@ -1689,6 +1694,11 @@ void admin_server::register_broker_routes() {
       ss::httpd::broker_json::stop_broker_maintenance,
       [this](std::unique_ptr<ss::httpd::request> req)
         -> ss::future<ss::json::json_return_type> {
+          if (!_controller->get_feature_table().local().is_active(
```
I am wondering if we should add feature check to `members_frontend` ? This way using maintenance mode would be controlled centrally not only at API level.

dotnwat replied:
yes. I'll do both. the one in members_frontend as a central operation, but at the http layer as well so we can provide better error messages than stuff like "operation not supported" or whatever error a missing rpc would cause.

dotnwat commented:
latest force-push:

* fixed merge conflict
* validate feature active in members frontend
* fixed bug with raft config overwriting member state

senior7515 commented:
> maintenance mode is persistent and stored in the controller. currently, only one node at a time may be in maintenance mode.

w/out reading the code, trying to understand the cover letter

what happens if you PUT/DELETE/PUT/DELETE in rapid succession. Does the controller put the node in an irreversible state until it finishes, or can it be "undone" at any time. If so, i assume it is a quorum write to do the PUT/DELETE/PUT/DELETE ops. 

(not sure if we have a test for a series of rapid succession, testing the locking mechanism)

dotnwat commented:
> what happens if you PUT/DELETE/PUT/DELETE in rapid succession

last state will win. we will end up deduplicating these states to avoid flapping, but the service will tolerate it well with fast replay.

senior7515 commented:
@coral-waters ^ great discussion to add to the docs.

BenPope reviewed (commented):
Aside from my confusion, this looks good.

I think a second review from somebody who knows the code a bit better would help.

coral-waters commented:
I created docs issue https://github.com/redpanda-data/documentation/issues/266

jcsp commented on tests/rptest/tests/maintenance_test.py:
```diff
@@ -0,0 +1,146 @@
+        try:
+            self._enable_maintenance(other)
+        except requests.exceptions.HTTPError as e:
+            if "invalid state transition" in e.response.text:
```
nit: could also assert the status code (it's nice to distinguish 500s and 400s especially)

dotnwat replied:
fixed

jcsp reviewed (commented):
This is very fly, and I can see us using that central maintenance mode status for other things in future.

Minor comments only.

dotnwat commented on src/v/cluster/members_table.cc:
```diff
@@ -70,6 +71,7 @@ void members_table::update_brokers(
                 // do not override membership state of brokers
                 it->second->set_membership_state(membership_state);
             }
+            it->second->set_maintenance_state(maintenance_state);
```
Oh I read it more than a few times haha. I think I'm more awake today and I'll add a couple commits to clean it up.

dotnwat replied:
fixed this with some extra commits to clean it up bit. also clarified the intended semantics with @mmaslankaprv

dotnwat commented on src/v/raft/consensus.h:
```diff
@@ -318,6 +318,7 @@ class consensus {
 
     std::vector<follower_metrics> get_follower_metrics() const;
     result<follower_metrics> get_follower_metrics(model::node_id) const;
+    bool has_followers() const { return _fstats.size() > 0; }
```
are you referring to a situation when r=3 is reconfigured to be r=1?

@mmaslankaprv maybe you know the answer to john's question? switching the source for has_followers should be easy enough if we need to.

mmaslankaprv replied:
we do trim `_fstats` every time configuration changes. This check looks good.

dotnwat replied:
thanks

dotnwat commented on src/v/cluster/drain_manager.cc:
```diff
@@ -0,0 +1,270 @@
+        }
+    }
+
+    vlog(clusterlog.info, "Node draining has completed");
```
good point, fixed.

the service is a little too shared-nothing and it looks like I need to have something like an all-to-all reduce if we wanted to have a definitive "node is done draining" status. I think should be straight forward to improve later on.

dotnwat commented on src/v/cluster/members_table.cc:
```diff
@@ -140,6 +142,56 @@ std::vector<model::node_id> members_table::get_decommissioned() const {
+    }
+
+    /*
+     * enforce one-node-at-a-time in maintenance mode rule
```
my mental model for the controller log is that its a classic replicated state machine in which the result of a command isn't known until replay. in this view, yes, incompatible commands can be in log and the error we are generating here is ultimately returned back to the client making the request. I don't think this would stop the replay of other message, but perhaps I'm missing something there. can you say more about that concern?

regarding validation up front: (1) at a minimum we could / should do validation up front to avoid unnecessary noise in the log, but (2) more than a naive check of in-memory state would be needed to maintain serializability. I think @rystsov has demonstrated this in some other area where we propagate the expected term down into the storage layer. I'm less clear how this mode of upfront validation and strong consistency work.

I'll go a head add the basic up front validation, tho, I think that makes a lot of sense.

dotnwat replied:
update on this I see we aren't doing any other validation in the members frontend so presumably we'd need to bring in a reference to the members table. what do you think @mmaslankaprv ?

mmaslankaprv replied:
i think basic validation (without using `term`) and linearizable barriers is just fine in the topics frontend. That would be an optimization. Final validation will be made during applying entries to STM.

dotnwat replied:
> i think basic validation (without using term) and linearizable barriers is just fine in the topics frontend. That would be an optimization. Final validation will be made during applying entries to STM.

are you saying the validation should be in topics frontend and not in the members_frontend? do you have an example pointer of a similar type of validation being done?

dotnwat replied:
@mmaslankaprv just to clarify, are you saying that creating topics is related to maintenance mode?

dotnwat replied:
Spoke with Michal offline. The current version in the PR is the easiest solution that provides strong consistency. Other validation can be done upfront with some more effort to avoid writing commands that are known to fail but are optimizations.

Will add this optimization along with the deduplication to avoid flapping of the service in a follow up PR.

dotnwat commented on src/v/redpanda/admin/api-doc/partition.json:
```diff
@@ -241,6 +241,10 @@
+                },
+                "is_leader": {
+                    "type": "boolean",
+                    "description": "node is leader"
```
ooph yeh that is a bit awkward.

I changed it from `is_leader` to `leader` with the description `Latest known leader`. Then the caller can reason about what that means. In this case it means comparing the latest known leader to the node in question and that can be done on the test side.

The only reason to do this through the admin api and not the kafka metadata api is I want to also know about the controller raft partition, and it avoids any messy out of date caching issues that might arise.

dotnwat commented on tests/rptest/services/admin.py:
```diff
@@ -358,3 +358,32 @@ def _has_leader():
+        url = f"brokers/{id}/maintenance"
+        self.redpanda.logger.info(
+            f"Starting maintenance on node {node.name}/{id}")
+        return self._request("put", url, node=target)
```
removed the target parameter

dotnwat commented on src/v/cluster/types.h:
```diff
@@ -839,6 +839,15 @@ struct finish_reallocation_reply {
     errc error;
 };
 
+struct set_maintenance_mode_request {
```
good call I shouldn't have missed this. fixed.

dotnwat commented:
https://github.com/redpanda-data/redpanda/compare/dd6028f14b440db90f53d171cc0ca048ce44153f..87903f8cf0cccc21fa588b3287d7b259746e8728

for the small diff

dotnwat commented:
Failure was https://github.com/redpanda-data/redpanda/issues/3924 fixed with integer overflow fix. Restarting that which should pick up the change in the CI rebase.

jcsp reviewed (approved):
No further questions from me!

BenPope reviewed (approved):
LGTM
